crates/python_extractor/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+            input_str.as_str(),
+        ));
+
+        if !disable_ref_generation {
+            let program = PythonProgram::parse(&input_str, "foo.py").with_context(|| {
+                format!(
+                    "Error while parsing file {:?
+        }",
+                    input_file
+                )
+            })?;
+            refs.extend(extract_py_imports::extract(&program));
+        }
+
+        let file_p = input_file.to_string_lossy();
+
+        if let Some(rel) = import_path_relative_from.as_ref() {
+            defs.extend(expand_path_to_defs_from_offset(rel, file_p.as_ref()));
+        } else {
+            defs.extend(expand_path_to_defs(file_p.as_ref()));
+        }
+        data_blocks.push(DataBlock {
+            entity_path: relative_path,
+            defs,
+            refs,
+            bzl_gen_build_commands,
+        })
+    }
+
+    let def_refs = ExtractedData {
+        label_or_repo_path: label_or_repo_path.clone(),
+        data_blocks,
+    };
+
+    tokio::fs::write(output, serde_json::to_string_pretty(&def_refs)?).await?;
+    Ok(())
+}
+
+fn expand_path_to_defs_from_offset(from_given_path: &str, path: &str) -> Vec<String> {
+    // rules_python Bzlmod support uses pip-tools, which I think places the 3rdparty
+    // source files inside a site-packages/ directory, per module.
+    if let Some(rem) = path
+        .strip_prefix(from_given_path)
+        .and_then(|p| Some(p.strip_prefix("site-packages/").unwrap_or(p)))
+    {
+        if let Some(e) = rem.strip_suffix(".py") {
+            let targ = e.replace('/', ".");
+
+            if let Some(p) = targ.strip_suffix(".__init__") {
+                return vec![p.to_string(), targ.clone()];
+            } else {
+                return vec![targ];
+            }
+        }
+    }
+    Vec::default()
+}
+
+fn expand_path_to_defs(path: &str) -> Vec<String> {
+    let mut results = Vec::default();
+    for element in path.split('/') {
+        results = results
+            .into_iter()
+            .map(|r| format!("{}.{}", r, element))
+            .collect();
+
+        if element == "src" {
+            results.push("src".to_string());
+        }
+    }
+
+    let mut results: Vec<String> = results
+        .into_iter()
+        .map(|e| e.strip_suffix(".py").map(|e| e.to_string()).unwrap_or(e))
+        .collect();
+
+    results.sort();
+    results.dedup();
+    results
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn expand_path_to_defs_test() {
+        let mut expected = vec!["src.main.python.blah.ppp"];
+        expected.sort();
+        expected.dedup();
+
+        assert_eq!(
+            expand_path_to_defs("/Users/foo/bar/src/main/python/blah/ppp.py"),
+            expected
+        );
+    }
+
+    #[test]
+    fn expand_path_to_defs_ambigious_path_test() {
+        let mut expected = vec!["src.main.python.blah.src.main.ppp", "src.main.ppp"];
+        expected.sort();
+        expected.dedup();
```

Comment:
why do we need this? That list doesn't have duplicates.

Reply:
this was written by someone else, but i will fix it 😉